### PR TITLE
Only disable following steps in the checkout process when the current step has a continue button

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | In the checkout process only disable following steps when the current step has a continue button This will prevent to block the workflow in themes which don't have the continue button in the "Personal information" step (when logged in)
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/14846
| How to test?  | First you need to be logged in with a PrestaShop frontend user, then in the checkout process when you are (at least) at the address step you can go backward by clicking on the "Personal information" title<br /><br />Now this bug only occurs if you have a theme which doesn't include the continue button, so to simulate this you need to edit the template `themes/classic/templates/checkout/_partials/steps/personal-information.tpl` and remove continue button (the one with class `continue`)<br /><br />When you refresh and retry the previous actions on checkout page you will see that when you click on "Personal Information" title, the next steps are clickable but as soon as you go to the next step you need to click on Continue<br />So this fix shouldn't break the required change from https://github.com/PrestaShop/PrestaShop/issues/13126

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/15042)

### CODE REVIEW

**Review by matthieu-rolland — DISMISSED** [review]
lgtm

**Comment by khouloudbelguith** [comment]
Hi @jolelievre,

In this case:
1. you are just logged in with a PrestaShop frontend user
2. Add some products to the cart
3. go to the checkout process
4. You are in "Addresses Step"
5. Click on the "Personal Information" step
6. The step "Addresses" is not disabled => OK but the edit button is not displayed
image

7. but, you cannot click on it => Error
8. if you reload the page => you are in the "Addresses Step" again
9. Click on Continue button => you are on the "Shipping Method" step
10. Click on the "Personal Information" step
11. The step "Addresses" is not disabled => OK & The Edit button is displayed
12. So you can click on the "Adress" Step
image

Thanks!

**Comment by jolelievre** [comment]
@khouloudbelguith 
thanks, seems like I didn't test enough the use case when address is not set yet. I'm gonna try it!

**Comment by jolelievre** [comment]
@khouloudbelguith 
Ok I fixed the bug, you can now click on Addresses when no "continue" button is present in the "Personal informations"

About the other steps you mentioned, when a step has already been completed and you click on continue it is logical that you move to the latest step, unless your modification requires for a change in another step (for example if you change your address you might need to redefine your shipping method)

**Comment by khouloudbelguith** [comment]
@jolelievre, thanks!
It is OK now.
Waiting for review

File: themes/_core/js/checkout.js [hunk 1]
```diff
@@ -51,9 +54,17 @@ function handleCheckoutStepChange() {
...
+          //If the step has no continue button, the previously completed steps are clickable
+          const $nextSteps = $clickedStep.nextAll('.checkout-step.-clickable');
+          $nextSteps.removeClass('-unreachable').addClass('-complete');
+          $('.step-title', $nextSteps).removeClass('not-allowed');
+        } else {
+          //If the step has a continue button, all next steps are disabled in order to force the user to click on continue
+          const $nextSteps = $clickedStep.nextAll();
+          $nextSteps.addClass('-unreachable').removeClass('-complete');
+          $('.step-title', $nextSteps).addClass('not-allowed');
+        }
```
Comment (eternoendless) [inline themes/_core/js/checkout.js:67]: Consider extracting functions and objects, this code is very complicated to understand.

Picture this:

```js
if (!clickedStep.isUnreachable()) {
	clickedStep.makeCurrent();
	if (clickedStep.hasContinueButton()) {
		steps.disableAllAfter(clickedStep)
	} else {
        steps.enableAllBefore(clickedStep)
    }
}
```

Looks better isn't it?

Comment (jolelievre) [reply themes/_core/js/checkout.js:67]: Yep much better, but then I need to extend jQuery functions? I'm not familiar with it but I can try
Or could this be enough:
```javascript
if (!isUnreachable(clickedStep)) {
	makeCurrent(clickedStep);
	if (hasContinueButton(clickedStep)) {
		disableAllAfter(steps, clickedStep)
	} else {
        	enableAllBefore(steps, clickedStep)
    	}
}

```

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Requested change from @eternoendless will be handled for 1.7.7
Current PR will be merged for 1.7.6.x as it has been validated ✅by @khouloudbelguith

**Pull request merged**